src/engine/renderer/tr_image_ktx.cpp

```diff
@@ -38,143 +41,191 @@ struct KTX_header_t {
+		  hdr->pixelWidth == 0 || hdr->pixelHeight == 0 ||
+		  hdr->pixelDepth != 0 );
+}
 
-	// Support only for RGBA8 and BCn
-	switch( hdr->glInternalFormat ) {
+static bool TryParseInternalFormatBits( uint32_t glInternalFormat, int &bits ) {
+	switch (glInternalFormat) {
 	case GL_RGBA8:
-		break;
+		return true;
 	case GL_COMPRESSED_RGB_S3TC_DXT1_EXT:
-		*bits |= IF_BC1;
-		break;
+		bits |= IF_BC1;
+		return true;
 	case GL_COMPRESSED_RGBA_S3TC_DXT3_EXT:
-		*bits |= IF_BC2;
-		break;
+		bits |= IF_BC2;
+		return true;
 	case GL_COMPRESSED_RGBA_S3TC_DXT5_EXT:
-		*bits |= IF_BC3;
-		break;
+		bits |= IF_BC3;
+		return true;
 	case GL_COMPRESSED_RED_RGTC1:
-		*bits |= IF_BC4;
-		break;
+		bits |= IF_BC4;
+		return true;
 	case GL_COMPRESSED_RG_RGTC2:
-		*bits |= IF_BC5;
-		break;
+		bits |= IF_BC5;
+		return true;
 	default:
-		ri.FS_FreeFile( hdr );
-		return;
+		return false;
 	}
+}
 
-	if( hdr->numberOfArrayElements != 0 ||
-	    (hdr->numberOfFaces != 1 && hdr->numberOfFaces != 6) ||
-	    hdr->pixelWidth == 0 ||
-	    hdr->pixelHeight == 0 ||
-	    hdr->pixelDepth != 0 ) {
-		ri.FS_FreeFile( hdr );
-		return;
+static uint32_t GetImageSize( byte *ptr, bool needReverseBytes ) {
+	uint32_t imageSize = *((uint32_t *)ptr);
+
+	if( needReverseBytes )
+		imageSize = Swap32(imageSize);
+
+	return PAD(imageSize, 4);
+}
+
+static bool IsNonArrayCubemapTexture( KTX_header_t *hdr ) {
+	return hdr->numberOfArrayElements == 0 && hdr->numberOfFaces == 6;
+}
+
+static bool LoadInMemoryKTX( const char *name, void *ktxData, size_t ktxSize,
+    			     byte **data, int *width, int *height, int *numLayers,
+    			     int *numMips, int *bits ) {
+	auto *hdr{ static_cast<KTX_header_t *>(ktxData) };
+
+	if( !IsValidKTXHeader( hdr, ktxSize ) ) {
+		Log::Warn("KTX texture '%s' is not in KTX format", name);
+		return false;
+	}
+
+	bool needReverseBytes{false};
+	if( !TryApplyKTXHeaderEndianness( hdr, needReverseBytes ) ) {
+		Log::Warn("KTX texture '%s' has unknown endianness value %u", name,
+		hdr->endianness);
+		return false;
+	}
+
+	if( !IsSupportedKTXFormat( hdr ) ) {
+		Log::Warn("KTX texture '%s' may be valid, but not supported", name);
+		return false;
+	}
+
+	if( !TryParseInternalFormatBits( hdr->glInternalFormat, *bits ) ) {
+		Log::Warn("KTX texture '%s' has unsupported glInternalFormat value %u",
+		name, hdr->glInternalFormat);
+		return false;
 	}
 
 	*width = hdr->pixelWidth;
 	*height = hdr->pixelHeight;
 	*numMips = hdr->numberOfMipmapLevels;
 	*numLayers = hdr->numberOfFaces == 6 ? 6 : 0;
 
-	ptr = (byte *)(hdr + 1);
-	ptr += hdr->bytesOfKeyValueData;
-	size = 0;
-	for(unsigned i = 0; i < hdr->numberOfMipmapLevels; i++ ) {
-		imageSize = *((uint32_t *)ptr);
-		if( hdr->endianness == KTX_endianness_reverse )
-			imageSize = Swap32( imageSize );
-		imageSize = PAD( imageSize, 4 );
-
-		size += imageSize * hdr->numberOfFaces;
-		ptr += 4 + imageSize;
+	byte *firstImageDataPtr{ (byte *)(hdr + 1) + hdr->bytesOfKeyValueData };
```

Comment:
There should be validation that this is inside the bounds of the file.

Reply:
Done, thanks!